staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -742,16 +762,133 @@ func validateCustomResourceDefinitionValidation(ctx context.Context, customResou
 	return allErrs
 }
 
+// unbounded uses nil to represent an unbounded cardinality value.
+var unbounded *uint64 = nil
+
+type costInfo struct {
+	// MaxCardinality represents a limit to the number of data elements that can exist for the current
+	// schema based on MaxProperties or MaxItems limits present on parent schemas, If all parent
+	// map and array schemas have MaxProperties or MaxItems limits declared MaxCardinality is
+	// an int pointer representing the product of these limits.  If least one parent map or list schema
+	// does not have a MaxProperties or MaxItems limits set, the MaxCardinality is nil, indicating
+	// that the parent schemas offer no bound to the number of times a data element for the current
+	// schema can exist.
+	MaxCardinality *uint64
+	CRDCost        *crdCost
+}
+
+type crdCost struct {
+	totalCost     uint64
+	mostExpensive []ruleCost
+}
+
+func (c *crdCost) observeExpressionCost(path *field.Path, cost uint64) {
+	if math.MaxUint64-c.totalCost < cost {
+		c.totalCost = math.MaxUint64
+	} else {
+		c.totalCost += cost
+	}
+
+	if cost < StaticEstimatedCRDCostLimit/100 { // ignore rules that contribute < 1% of total cost
+		return
+	}
+	c.mostExpensive = append(c.mostExpensive, ruleCost{path: path, cost: cost})
+	sort.Slice(c.mostExpensive, func(i, j int) bool {
```

Comment:
Alternatively, I could do pure insertion sort on this (bubble in each new entry). For a size 4 list, I suspect it won't matter.